src/main/java/org/jenkinsci/plugins/stashNotifier/StashNotifier.java

```diff
@@ -636,21 +672,50 @@ private HttpEntity newStashBuildNotificationEntity(
         json.put("description", getBuildDescription(build, state));
         json.put("url", Jenkins.getInstance()
         		.getRootUrl().concat(build.getUrl()));
-        
+
         return new StringEntity(json.toString(), "UTF-8");
 	}
 
+
+
 	/**
-	 * Returns the build key used in the Stash notification. Includes the 
+	 * Returns the build key used in the Stash notification. Includes the
 	 * build number depending on the user setting.
-	 * 
+	 *
 	 * @param 	build	the build to notify Stash of
 	 * @return	the build key for the Stash notification
 	 */
-	private String getBuildKey(final AbstractBuild<?, ?> build) {
-		StringBuilder key = new StringBuilder();
+	private String getBuildKey(final AbstractBuild<?, ?> build,
+							   BuildListener listener) {
+
+        String overriddenKey = (projectKey != null && projectKey.trim().length() > 0) ? projectKey : getDescriptor().getProjectKey();
+
+		if (overriddenKey != null && overriddenKey.trim().length() > 0) {
+			PrintStream logger = listener.getLogger();
+			try {
+				EnvVars environment = build.getEnvironment(listener);
+				return StringEscapeUtils.escapeJavaScript(environment.expand(projectKey));
+			} catch (IOException e) {
+				logger.println("Unable to expand commit SHA value");
+				e.printStackTrace(logger);
+				return null;
```

Comment:
if you return `null` here, you would set `null` as the value of the JSON key `key` in line 662 above, which doesn't check for a null value returned by this method. Instead, I propose to log an error message like you do but instead use the project name as a default key.